Stream sources are not being freed after adding an RTSP endpoint

Relevant log:

```log
2023-01-11T20:04:42.154284Z TRACE PipelineRunner-9c4fcb3a-6075-4e29-90a6-b9065bc6f3ea ThreadId(51) runner{pipeline_weak=WeakRef(GWeakRef @ 0x7f11f0047bb0 { priv_: GWeakRef_priv @ 0x7f11f0047bb0 { p: 0x55603d7e1a40 } }, PhantomData<*mut gstreamer::auto::pipeline::Pipeline>) pipeline_id=9c4fcb3a-6075-4e29-90a6-b9065bc6f3ea killswitch_receiver=broadcast::Receiver}: src/stream/pipeline/runner.rs:152: StreamStart(
    StreamStart(
        Message {
            ptr: 0x00007f11fc0aab30,
            type: "stream-start",
            seqnum: 322,
            src: Some(
                "pipeline3",
            ),
            structure: Some(
                GstMessageStreamStart {
                    group-id: (guint) 4,
                },
            ),
        },
    ),
)
2023-01-11T20:04:42.154319Z ERROR PipelineRunner-9c4fcb3a-6075-4e29-90a6-b9065bc6f3ea ThreadId(51) runner{pipeline_weak=WeakRef(GWeakRef @ 0x7f11f0047bb0 { priv_: GWeakRef_priv @ 0x7f11f0047bb0 { p: 0x55603d7e1a40 } }, PhantomData<*mut gstreamer::auto::pipeline::Pipeline>) pipeline_id=9c4fcb3a-6075-4e29-90a6-b9065bc6f3ea killswitch_receiver=broadcast::Receiver}: src/stream/pipeline/runner.rs:131: Error from Some("/GstPipeline:pipeline3/GstV4l2Src:v4l2src3"): Device '/dev/video2' is busy (Some("../gstreamer/subprojects/gst-plugins-good/sys/v4l2/gstv4l2object.c(4053): gst_v4l2_object_set_format_full (): /GstPipeline:pipeline3/GstV4l2Src:v4l2src3:\nCall to S_FMT failed for H264 @ 1920x1080: Device or resource busy"))
2023-01-11T20:04:42.155220Z ERROR PipelineRunner-9c4fcb3a-6075-4e29-90a6-b9065bc6f3ea ThreadId(51) src/stream/pipeline/runner.rs:34: PipelineWatcher ended with error: Device '/dev/video2' is busy
2023-01-11T20:04:42.155244Z  INFO PipelineRunner-9c4fcb3a-6075-4e29-90a6-b9065bc6f3ea ThreadId(51) src/stream/pipeline/runner.rs:44: Error sent to killswitch channel!
2023-01-11T20:04:43.054200Z DEBUG ThreadId(52) src/stream/pipeline/mod.rs:134: Killswitch received as 9c4fcb3a-6075-4e29-90a6-b9065bc6f3ea from PipelineState's watcher. Reason: "Device '/dev/video2' is busy"
2023-01-11T20:04:43.054250Z  WARN ThreadId(52) src/stream/pipeline/mod.rs:137: Failed removing Pipeline 9c4fcb3a-6075-4e29-90a6-b9065bc6f3ea. Reason: Already removed

```

--- a/src/stream/manager.rs
+++ b/src/stream/manager.rs
@@ -185,7 +185,7 @@ pub fn add_stream_and_start(video_and_stream_information: VideoAndStreamInformat
     for stream in manager.streams.values() {
         stream
             .video_and_stream_information
-            .conflicts_with(&video_and_stream_information)?
+            .conflicts_with(&video_and_stream_information)?;
     }
     drop(manager);
     Manager::add_stream(stream)?;

--- a/src/stream/manager.rs
+++ b/src/stream/manager.rs
@@ -17,7 +17,7 @@ use crate::{
 
 use anyhow::{anyhow, Context, Result};
 
-use gst::{event::Eos, prelude::ElementExtManual, traits::ElementExt};
+use gst::{prelude::ElementExtManual, traits::ElementExt};
 use tracing::*;
 
 use super::{
@@ -354,7 +354,7 @@ impl StreamManagementInterface<StreamStatus> for Manager {
             return Err(anyhow!("Already removed"));
         }
 
-        let stream = manager
+        let mut stream = manager
             .streams
             .remove(stream_id)
             .context(format!("Stream {stream_id:?} not found"))?;
@@ -363,7 +363,9 @@ impl StreamManagementInterface<StreamStatus> for Manager {
 
         let pipeline = &stream.pipeline.inner_state_as_ref().pipeline;
         let pipeline_id = stream_id;
-        pipeline.send_event(Eos::new());
+        if let Err(error) = pipeline.post_message(gst::message::Eos::new()) {
+            error!("Failed posting Eos message into Pipeline' bus. Reason: {error:?}");
+        }
 
         if let Err(error) = stream
             .pipeline

--- a/src/stream/manager.rs
+++ b/src/stream/manager.rs
@@ -377,20 +377,21 @@ impl StreamManagementInterface<StreamStatus> for Manager {
             std::thread::sleep(std::time::Duration::from_millis(100));
         }
 
-        // Unlink all Sinks
-        stream
+        // Remove all Sinks
+        let sink_ids = &stream
             .pipeline
             .inner_state_as_ref()
             .sinks
-            .values()
-            .for_each(|sink| {
-                if let Err(error) = sink.unlink(pipeline, pipeline_id) {
-                    warn!(
-                        "Failed unlinking Sink {:?} from Pipeline {pipeline_id:?}. Reason: {error:?}",
-                        sink.get_id()
-                    );
-                }
-            });
+            .keys()
+            .cloned()
+            .collect::<Vec<uuid::Uuid>>();
+        for sink_id in sink_ids {
+            if let Err(error) = stream.pipeline.remove_sink(sink_id) {
+                warn!(
+                    "Failed unlinking Sink {sink_id:?} from Pipeline {pipeline_id:?}. Reason: {error:?}"
+                );
+            }
+        }
 
         info!("Stream {stream_id} successfully removed!");
 

--- a/src/stream/mod.rs
+++ b/src/stream/mod.rs
@@ -21,6 +21,8 @@ use anyhow::{anyhow, Result};
 
 use tracing::*;
 
+use self::rtsp::rtsp_server::RTSP_SERVER_PORT;
+
 #[derive(Debug)]
 pub struct Stream {
     pub id: PeerId,
@@ -140,6 +142,11 @@ fn validate_endpoints(video_and_stream_information: &VideoAndStreamInformation) 
                 if endpoint.host().is_none() || endpoint.port().is_none() || endpoint.path().is_empty() {
                     return Some(anyhow!(
                         "Endpoint with rtsp scheme should contain host, port, and path. Endpoint: {endpoint:?}"
+                    ));
+                }
+                if endpoint.port() != Some(RTSP_SERVER_PORT) {
+                    return Some(anyhow!(
+                        "Endpoint with rtsp scheme should use port {RTSP_SERVER_PORT:?}. Endpoint: {endpoint:?}"
                     ));
                 }
             }

--- a/src/stream/pipeline/runner.rs
+++ b/src/stream/pipeline/runner.rs
@@ -170,7 +170,7 @@ impl Drop for PipelineRunner {
             .killswitch_sender
             .send("PipelineRunner Dropped.".to_string())
         {
-            error!(
+            warn!(
                 "Failed to send killswitch message while Dropping PipelineRunner. Reason: {reason:#?}"
             );
         }

--- a/src/stream/rtsp/rtsp_server.rs
+++ b/src/stream/rtsp/rtsp_server.rs
@@ -33,10 +33,16 @@ impl RTSPServer {
         let is_running = false;
         let (sender, receiver) = std::sync::mpsc::channel::<String>();
 
+        let host = "0.0.0.0".to_string();
+        let port = RTSP_SERVER_PORT;
+        let server = gst_rtsp_server::RTSPServer::new();
+        server.set_address(&host);
+        server.set_service(&port.to_string());
+
         RTSPServer {
-            server: gst_rtsp_server::RTSPServer::new(),
-            host: "0.0.0.0".into(),
-            port: 8554,
+            server,
+            host,
+            port,
             run: is_running,
             path_to_factory: HashMap::new(),
             main_loop_thread: Some(
@@ -186,8 +192,6 @@ impl RTSPServer {
 
     #[instrument(level = "debug")]
     pub fn start_pipeline(path: &str) -> Result<()> {
-        RTSPServer::configure("0.0.0.0", 8554);
-
         let mut rtsp_server = RTSP_SERVER.as_ref().lock().unwrap();
 
         // Much like HTTP servers, RTSP servers have multiple endpoints that
@@ -231,21 +235,5 @@ impl RTSPServer {
         rtsp_server.path_to_factory.remove(path);
 
         Ok(())
-    }
-
-    #[instrument(level = "debug")]
-    fn configure(host: &str, port: u16) {
-        let mut rtsp_server = RTSP_SERVER.as_ref().lock().unwrap();
-        if rtsp_server.run {
-            return;
-        }
-        rtsp_server.host = host.to_string();
-        rtsp_server.port = port;
-
-        rtsp_server.server.set_address(&rtsp_server.host);
-
-        rtsp_server
-            .server
-            .set_service(&rtsp_server.port.to_string());
     }
 }

--- a/src/stream/rtsp/rtsp_server.rs
+++ b/src/stream/rtsp/rtsp_server.rs
@@ -23,6 +23,8 @@ lazy_static! {
         Arc::new(Mutex::new(RTSPServer::default()));
 }
 
+pub const RTSP_SERVER_PORT: u16 = 8554;
+
 impl RTSPServer {
     #[instrument(level = "debug")]
     fn default() -> Self {

--- a/src/stream/rtsp/rtsp_server.rs
+++ b/src/stream/rtsp/rtsp_server.rs
@@ -73,7 +73,10 @@ impl RTSPServer {
             // interested in them. In this example, we only do have one, so we can
             // leave the context parameter empty, it will automatically select
             // the default one.
-            let id = rtsp_server.server.attach(None).unwrap();
+            let id = rtsp_server
+                .server
+                .attach(None)
+                .expect("Failed attaching to default context");
 
             // Start the mainloop. From this point on, the server will start to serve
             // our quality content to connecting clients.
@@ -106,7 +109,7 @@ impl RTSPServer {
             .find_map(|structure| {
                 structure.iter().find_map(|(key, sendvalue)| {
                     if key == "encoding-name" {
-                        Some(sendvalue.to_value().get::<String>().unwrap())
+                        Some(sendvalue.to_value().get::<String>().expect("Failed accessing encoding-name parameter"))
                     } else {
                         None
                     }
@@ -191,7 +194,9 @@ impl RTSPServer {
             .mount_points()
             .context("Could not get mount points")?;
 
-        let factory = rtsp_server.path_to_factory.get(path).unwrap();
+        let factory = rtsp_server.path_to_factory.get(path).context(format!(
+            "Factory for path {path:?} not found in RTSP factories"
+        ))?;
 
         // Now we add a new mount-point and tell the RTSP server to serve the content
         // provided by the factory we configured above, when a client connects to

--- a/src/stream/rtsp/rtsp_server.rs
+++ b/src/stream/rtsp/rtsp_server.rs
@@ -24,6 +24,7 @@ lazy_static! {
 }
 
 impl RTSPServer {
+    #[instrument(level = "debug")]
     fn default() -> Self {
         if let Err(error) = gst::init() {
             error!("Failed to init GStreamer: {error:?}");
@@ -48,10 +49,12 @@ impl RTSPServer {
         }
     }
 
+    #[instrument(level = "debug")]
     pub fn is_running() -> bool {
         RTSP_SERVER.as_ref().lock().unwrap().run
     }
 
+    #[instrument(level = "debug")]
     fn run_main_loop(channel: std::sync::mpsc::Sender<String>) {
         if let Err(error) = gst::init() {
             let _ = channel.send(format!("Failed to init GStreamer: {error:?}"));
@@ -93,6 +96,7 @@ impl RTSPServer {
         }
     }
 
+    #[instrument(level = "debug")]
     pub fn add_pipeline(path: &str, socket_path: &str, rtp_caps: &gst::Caps) -> Result<()> {
         // Initialize the singleton before calling gst factory
         let mut rtsp_server = RTSP_SERVER.as_ref().lock().unwrap();
@@ -180,6 +184,7 @@ impl RTSPServer {
         Ok(())
     }
 
+    #[instrument(level = "debug")]
     pub fn start_pipeline(path: &str) -> Result<()> {
         RTSPServer::configure("0.0.0.0", 8554);
 
@@ -208,6 +213,7 @@ impl RTSPServer {
         Ok(())
     }
 
+    #[instrument(level = "debug")]
     pub fn stop_pipeline(path: &str) -> Result<()> {
         let mut rtsp_server = RTSP_SERVER.as_ref().lock().unwrap();
 
@@ -227,6 +233,7 @@ impl RTSPServer {
         Ok(())
     }
 
+    #[instrument(level = "debug")]
     fn configure(host: &str, port: u16) {
         let mut rtsp_server = RTSP_SERVER.as_ref().lock().unwrap();
         if rtsp_server.run {

--- a/src/stream/sink/rtsp_sink.rs
+++ b/src/stream/sink/rtsp_sink.rs
@@ -126,6 +126,7 @@ impl SinkInterface for RtspSink {
         self.sink_id
     }
 
+    #[instrument(level = "trace")]
     fn get_sdp(&self) -> Result<gst_sdp::SDPMessage> {
         Err(anyhow!(
             "Not available. Reason: RTSP Sink should only be connected from its RTSP endpoint."
@@ -170,10 +171,12 @@ impl RtspSink {
         })
     }
 
+    #[instrument(level = "trace")]
     pub fn path(&self) -> String {
         self.path.clone()
     }
 
+    #[instrument(level = "trace")]
     pub fn socket_path(&self) -> String {
         self.socket_path.clone()
     }

--- a/src/video_stream/types.rs
+++ b/src/video_stream/types.rs
@@ -6,7 +6,8 @@ use serde::{Deserialize, Serialize};
 
 use std::collections::HashSet;
 
-use simple_error::{simple_error, SimpleResult};
+use anyhow::{anyhow, Result};
+
 //TODO: move to stream ?
 #[derive(Apiv2Schema, Clone, Debug, PartialEq, Deserialize, Serialize)]
 pub struct VideoAndStreamInformation {
@@ -16,24 +17,24 @@ pub struct VideoAndStreamInformation {
 }
 
 impl VideoAndStreamInformation {
-    pub fn conflicts_with(&self, other: &VideoAndStreamInformation) -> SimpleResult<()> {
+    pub fn conflicts_with(&self, other: &VideoAndStreamInformation) -> Result<()> {
         if self.name == other.name {
-            return Err(simple_error!(format!(
+            return Err(anyhow!(
                 "Stream ({other_name:#?} - {other_source:#?}) is already using the name {name:#?}.",
                 other_name = other.name,
                 other_source = other.video_source.inner().source_string(),
                 name = self.name,
-            )));
+            ));
         }
 
         if (!self.video_source.inner().is_shareable())
             && (self.video_source.inner().source_string()
                 == other.video_source.inner().source_string())
         {
-            return Err(simple_error!(format!(
+            return Err(anyhow!(
                 "Streams have same source: {:#?}",
                 self.video_source.inner().source_string()
-            )));
+            ));
         }
 
         let our_endpoints: HashSet<_> = self.stream_information.endpoints.iter().collect();
@@ -41,14 +42,14 @@ impl VideoAndStreamInformation {
         let common_endpoints: HashSet<_> = our_endpoints.intersection(&other_endpoints).collect();
 
         if !common_endpoints.is_empty() {
-            return Err(simple_error!(format!(
+            return Err(anyhow!(
                 "Stream ({other_name:#?} - {other_source:#?}) has common endpoint with Stream ({our_name:#?} - {our_source:#?}). The common endpoint: {endpoints:#?}",
                 other_name = other.name,
                 other_source = other.video_source.inner().source_string(),
                 our_name = self.name,
                 our_source = self.video_source.inner().source_string(),
                 endpoints = common_endpoints
-            )));
+            ));
         }
 
         Ok(())

